backend/remote-state/http/client.go

```diff
@@ -0,0 +1,298 @@
+		}
+		return "", c.lockError(fmt.Errorf("HTTP remote state already locked: ID=%s", existing.ID))
+	default:
+		return "", fmt.Errorf("Unexpected HTTP response code %d", resp.StatusCode)
+	}
+}
+
+// Unlock the state with id from lock.
+func (c *RemoteClient) Unlock(id string) error {
+
+	lockErr := &state.LockError{}
+
+	lockInfo, err := c.lockInfo()
+	if err != nil {
+		lockErr.Err = fmt.Errorf("failed to retrieve lock info: %s", err)
+		return lockErr
+	}
+	lockErr.Info = lockInfo
+
+	if lockInfo.ID != id {
+		lockErr.Err = fmt.Errorf("lock id %q does not match existing lock", id)
+		return lockErr
+	}
+
+	unlockURL, _ := url.Parse(c.unlockAddress)
+	resp, err := c.httpRequest(c.unlockMethod, unlockURL, &c.jsonLockInfo, "unlock")
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		return nil
+	default:
+		return fmt.Errorf("Unexpected HTTP response code %d", resp.StatusCode)
+	}
+}
+
+func (c *RemoteClient) lockError(err error) *state.LockError {
+	lockErr := &state.LockError{
+		Err: err,
+	}
+
+	info, infoErr := c.lockInfo()
+	if infoErr != nil {
+		lockErr.Err = multierror.Append(lockErr.Err, infoErr)
+	} else {
+		lockErr.Info = info
+	}
+	return lockErr
+}
+
+// lockInfo reads the lock file, parses its contents and returns the parsed
+// LockInfo struct.
+func (c *RemoteClient) lockInfo() (*state.LockInfo, error) {
+	// Convert address to type URL
+	lockURL, _ := url.Parse(c.lockAddress)
+
+	resp, err := c.httpRequest("GET", lockURL, nil, "get lock")
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	// Read in the body
+	buf := bytes.NewBuffer(nil)
+	if _, err := io.Copy(buf, resp.Body); err != nil {
+		return nil, fmt.Errorf("Failed to read lock file: %s", err)
+	}
+
+	info := &state.LockInfo{}
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		readBuf, _ := ioutil.ReadAll(buf)
+		if err := json.Unmarshal(readBuf, info); err != nil {
+			return nil, err
+		}
+		return info, nil
+	default:
+		return nil, fmt.Errorf("Unexpected HTTP response code %d", resp.StatusCode)
+	}
+}
+
+func (c *RemoteClient) httpRequest(method string, url *url.URL, data *[]byte, what string) (*http.Response, error) {
+	// If we have data we need a reader
+	var reader io.Reader
+	if data != nil {
+		reader = bytes.NewReader(*data)
+	}
+	// Create the request
+	req, err := http.NewRequest(method, url.String(), reader)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to make %s HTTP request: %s", what, err)
+	}
+	// Setup basic auth
+	if c.username != "" {
+		req.SetBasicAuth(c.username, c.password)
+	}
+
+	// Work with data/body
+	if data != nil {
+		req.Header.Set("Content-Type", "application/json")
+		req.ContentLength = int64(len(*data))
+
+		// Generate the MD5
+		hash := md5.Sum(*data)
+		b64 := base64.StdEncoding.EncodeToString(hash[:])
+		req.Header.Set("Content-MD5", b64)
+	}
+
+	// Make the request
+	resp, err := c.client.Do(req)
```

Comment:
You can condense all this to `return c.Client.Do(req)`. You don't need to check if there was an error, and then return an error? Just bubble up the error from this call.